service/storage_proxy.cc

```diff
@@ -3525,6 +3559,9 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
     }
     future<> make_digest_requests(digest_resolver_ptr resolver, targets_iterator begin, targets_iterator end, clock_type::time_point timeout) {
         return parallel_for_each(begin, end, [this, resolver = std::move(resolver), timeout] (gms::inet_address ep) {
+            if (maybe_speculatively_fail(*resolver, ep, timeout)) {
```

Comment:
Why speculatively fail if we can speculatively choose another replica when we calculate targets?

Reply:
We would choose another replica via speculative retry (if configured), which has the advantage that the mechanism already exists and works. The intention is that Scylla would behave exactly as if this target node was already dead/overloaded, but without actually allocating a background read for it, thus saving coordinator's resources.

 Initially I started this series by implementing this mechanism at the point where targets are picked as you suggested, but the change was rather intrusive and I gave it up after coming up with more and more corner cases to consider. Still, perhaps the complexity is worth the effort, since we'd pick another node (if possible) right away, instead of introducing a delay by relying on speculative retry. I'll take another shot at introducing this mechanism sooner, perhaps I changed my mind to soon.